src/vg_active_segment.erl

```diff
@@ -99,67 +101,127 @@ init([Topic, Partition, NextNode, Tab]) ->
-                byte_count=0,
-                pos=Position,
-                index_pos=IndexPosition,
-                log_fd=LogFD,
-                segment_id=list_to_integer(LastLogId),
-                index_fd=IndexFD,
-                topic=Topic,
-                partition=Partition,
-                config=Config,
-                history_tab=Tab
+    {ok, #state{id = Id,
+                next_brick = {?SERVER(Topic, Partition), NextNode},
+                topic_dir = TopicDir,
+                byte_count = 0,
+                pos = Position,
+                index_pos = IndexPosition,
+                log_fd = LogFD,
+                segment_id = list_to_integer(LastLogId),
+                index_fd = IndexFD,
+                topic = Topic,
+                partition = Partition,
+                config = Config,
+                history_tab = Tab
                }}.
 
-handle_call({write, RecordSet}, _From, State=#state{id=Id,
-                                                    topic=Topic,
-                                                    partition=Partition,
-                                                    next_brick=NextBrick}) ->
-    State1 = write_record_set(RecordSet, State),
-    case NextBrick of
-        {_, solo} -> ok;
-        {_, tail} -> ok;
-        {_, last} -> ok;
-        _ ->
-            gen_server:cast(NextBrick, {write, self(), RecordSet})
-    end,
-    LatestId = State1#state.id,
-    %% Should we do this in ack instead if it isn't solo or tail?
-    vg_topics:update_hwm(Topic, Partition, LatestId-1),
-    {reply, {ok, Id}, State1};
+handle_call({write, ExpectedID0, RecordSet}, _From, State = #state{id = ID,
+                                                                   topic = Topic,
+                                                                   partition = Partition,
+                                                                   next_brick = NextBrick}) ->
+    %% TODO: add pipelining of requests
+    try
+        ExpectedID =
+            case ExpectedID0 of
+                head ->
+                    ID + length(RecordSet);
+                Supplied when is_integer(Supplied) ->
+                    case ID + length(RecordSet) == Supplied of
+                        true ->
+                            ExpectedID0;
+                        %% should we check > vs < here?  one is repair
+                        %% the other is bad corruption
+                        _ ->
+                            %% inferred current id of the writing segment
+                            WriterID = ExpectedID0 - length(RecordSet),
+                            %% this should probably be limited, if
+                            %% we're going back too far, we need to be
+                            %% in some sort of catch-up mode
+                            lager:info("starting write repair, ~p", [WriterID]),
+                            WriteRepairSet = write_repair(WriterID, State),
```

Comment:
Add a prometheus metric in here so we can monitor the amount of repairs. Probably want one counter for just all the times a write repair is needed and a gauge for the size of the set.